Add LangChain example (#1044)
neondatabase/website: content/docs/ai/ai-langchain.md

---
title: Use LangChain with Neon
subtitle: Chat with your Neon database using LangChain and OpenAI
enableTableOfContents: true
---

> LangChain is an open-source framework for developing applications powered by Large Language Models (LLMs). It enables developers to build applications that use LLMs to reason about how to provide answers and take actions based on the provided context, which includes prompt instructions, example inputs and responses, and the data that responses should be based on.

Reviewer: ```suggestion
LangChain is an open-source framework for developing applications powered by Large Language Models (LLMs). Using the LangChain framework, you can build applications that offer answers or take actions in response to a variety of possible inputs or contexts: prompt instructions, example inputs or responses, or any other data that the reactions should be based on.
```

> For example, with LangChain, you can connect LLMs with an data source such as Neon Postgres and sequence multiple commands (a series of prompts) to achieve a desired result.

Reviewer: Typo an -> a:

...connect to a data source...

> For example, with LangChain, you can connect LLMs with an data source such as Neon Postgres and sequence multiple commands (a series of prompts) to achieve a desired result.

Reviewer: ```suggestion
For example, with LangChain, you can connect LLMs with a data source such as Neon Postgres and sequence multiple commands (a series of prompts) to achieve a desired result.
```

> In this guide, you will learn how to create a Python application that allows you to ask your database questions using natural language. The application transforms your question into an SQL query, runs the query, and returns a natural language response based on the query result. The application uses the [Titanic Dataset](https://www.kaggle.com/datasets/vinicius150987/titanic3). When you have the application up and running, you will be able to ask your database questions about the passengers of the Titanic.

Reviewer: ```suggestion
In this guide, you will learn how to create a Python application that lets you to ask your database questions using natural language. The application transforms your question into an SQL query, runs the query, and then returns a natural language response based on the query result. The application uses the [Titanic Dataset](https://www.kaggle.com/datasets/vinicius150987/titanic3) as the data source. Once you have the application up and running, you can use it to ask your database detailed questions about the passengers of the Titanic.
```

## Create a directory for your application

```bash
mkdir titanic_app
cd titanic_app
```

## Set up your environment

Create and activate a Python virtual environment in your application directory by executing the following command:

```python
python -m venv venv
```

## Install dependencies

Install `LangChain`, `openai`, `python-environ`, `psycopg2`, `environ`, and `requests` libraries using `pip`.

```sql
pip install LangChain LangChain-experimental openai python-environ psycopg2 
```

## Create a database in Neon

In Neon, create a database named `titanic`. You can do this from the Neon Console:

1. Navigate to the [Neon Console](https://console.neon.tech).
1. Select a project.
1. Select **Databases**.
1. Select the branch where you want to create the database.
1. Click **New Database**.
1. Enter `titanic` as the database name, and select a database owner.
1. Click **Create**.

## Retrieve your database connection string

Retrieve the connection string for your `titanic` database. You can copy it from the **Connection Details** widget on the Neon **Dashboard**. It will look something like this:

```text

> postgres://<users>:<password>@ep-morning-limit-06265452.us-east-2.aws.neon.tech/titanic

Reviewer: ```suggestion
postgres://[users]:[password]@ep-morning-limit-06265452.us-east-2.aws.neon.tech/titanic
```

```

## Set up your environment variables

Create a `.env` file in your application directory and configure variables for your OpenAI API key and database connection string:

```text

> OPENAI_API_KEY=<your_openai_api_key>

Reviewer: ```suggestion
OPENAI_API_KEY=[your_openai_api_key]
```

> DATABASE_URL="postgres://<users>:<password>@ep-morning-limit-06265452.us-east-2.aws.neon.tech/titanic"

Reviewer: ```suggestion
DATABASE_URL="postgres://[users]:[password]@ep-morning-limit-06265452.us-east-2.aws.neon.tech/titanic"
```

from langchain.prompts import PromptTemplate

# Set up and read the .env file
env = environ.Env()
environ.Env.read_env()

# Extract connection details from the .env file
connection_string = env('DATABASE_URL')
parsed_uri = urlparse(connection_string)
username = parsed_uri.username
password = parsed_uri.password
host = parsed_uri.hostname
port = parsed_uri.port or 5432
database = parsed_uri.path[1:]  # remove leading '/'

# Setup database
db = SQLDatabase.from_uri(
    f"postgresql+psycopg2://{username}:{password}@{host}:{port}/{database}"
)

# Setup llm
llm = OpenAI(temperature=0, openai_api_key=os.environ["OPENAI_API_KEY"])

# Define table_info and few_shot_examples
table_info = """public.passenger (
    passengerid integer NOT NULL,
    survived double precision,
    pclass integer,
    name text,
    sex text,
    age double precision,
    sibsp integer,
    parch integer,
    ticket text,
    fare double precision,
    cabin text,
    embarked text,
    wikiid double precision,
    name_wiki text,
    age_wiki double precision,
    hometown text,
    boarded text,
    destination text,
    lifeboat text,
    body text,
    class integer
)"""

> few_shot_examples = """

Reviewer: Should this read few_short_example?

Author: This one is OK. That's how OpenAI defines it: "few_shot_examples": https://python.langchain.com/docs/use_cases/qa_structured/sql

  - `psycopg2`: The PostgreSQL adapter for Python, allowing interaction with the PostgreSQL database.
  - `environ`: Used to manage environment variables and `.env` files.
  - `urlparse`: Part of Python's standard library for parsing URLs.
  - `langchain` related imports: A suite of utilities and modules for querying the database using natural language.

- **Setup and Read the .env File**:
  - The `environ` module is used to read environment variables, specifically the database connection details, from a `.env` file.

- **Extract Connection Details from the .env File**:
  - The connection string is retrieved from the `.env` file and parsed to extract database connection details like username, password, host, port, and database name.

- **Setup Database Connection**:
  - The `SQLDatabase.from_uri()` method from the `langchain.utilities` module establishes a connection to the PostgreSQL database using the extracted details.

- **Setup Language Learning Model (llm)**:
  - An instance of `OpenAI` from `langchain.llms` is created, with certain parameters like temperature. It also requires an API key, which is fetched from the environment variables.

> - **Defining Table Information and Few Shot Examples**:

Reviewer: Should the term be hyphenated: Few-Shot 
?

Author: This one is OK. This is how it's defined in the OpenAI docs. It's their term: https://python.langchain.com/docs/use_cases/qa_structured/sql

- **Defining Custom Prompt**:
  - The `TEMPLATE` provides a format that the language model should adhere to when processing input questions. 
  - `CUSTOM_PROMPT` is an instance of `PromptTemplate` that formats the TEMPLATE using the provided input variables.

- **Setting up the Database Chain**:
  - `SQLDatabaseChain` initializes the chain between the language learning model and the SQL database. This chain is responsible for processing the user's questions and fetching answers from the database.

- **Interactive Prompt Function**:
  - The `get_prompt()` function initiates an interactive session for the user to ask questions. 
  - For each question, the app tries to formulate an SQL query using the language model, executes the query on the database, and returns the answer.
  - If the user types "exit", the loop breaks, and the session ends.

- **Start the Interactive Session**:
  - The last line calls the `get_prompt()` function, starting the interactive session when the script is run.

</details>

## Run the application

> Run the application using the following command. When prompted, ask your `titanic` database a question like, "How many passengers survived?" or "What was the average age of passengers?".

Reviewer: Consider moving "When prompted... age of passengers?"> sentence below the bash command.

Author: Thank you. Moved, as suggested,
